internal/dialer/dialer.go

```diff
@@ -0,0 +1,256 @@
+		address = net.JoinHostPort(address, port)
+		conn, err := d.Connector.DialContext(ctx, network, address)
+		if err == nil {
+			return conn, nil
+		}
+		errDial.Errors = append(errDial.Errors, err)
+	}
+	return nil, errDial
+}
+
+// ErrDial indicates that DialContext failed
+type ErrDial struct {
+	// Errors contains the error of each connect() that failed.
+	Errors []error
+}
+
+// Error implements error.Error
+func (ErrDial) Error() string {
+	return "connect_error"
+}
+
+// Logger is the logger interface assumed by this package
+type Logger interface {
+	Debugf(format string, v ...interface{})
+}
+
+// LoggingDialer is a dialer that implements logging.
+type LoggingDialer struct {
+	Dialer
+	Logger Logger
+}
+
+// DialContext implements Dialer.DialContext.
+func (d LoggingDialer) DialContext(
+	ctx context.Context, network, address string) (net.Conn, error) {
+	d.Logger.Debugf("dial %s/%s...", address, network)
+	conn, err := d.Dialer.DialContext(ctx, network, address)
+	d.Logger.Debugf("dial %s/%s... %+v", address, network, err)
+	return conn, err
+}
+
+// ErrWrapper is a dialer that wraps errors
+type ErrWrapper struct {
+	Dialer
+}
+
+// DialContext implements Dialer.DialContext.
+func (d ErrWrapper) DialContext(
+	ctx context.Context, network, address string) (net.Conn, error) {
+	conn, err := d.Dialer.DialContext(ctx, network, address)
+	err = errwrapper.SafeErrWrapperBuilder{
+		Error:     err,
+		Operation: "connect",
+	}.MaybeBuild()
+	return conn, err
+}
+
+// EventsSaver is a dialer that saves events
+type EventsSaver struct {
+	Dialer
+	events []Events
+	mu     sync.Mutex
+}
+
+// ReadEvents reads the saved events and returns them
+func (d *EventsSaver) ReadEvents() []Events {
+	d.mu.Lock()
+	ev := d.events
+	d.events = nil
+	d.mu.Unlock()
+	return ev
+}
+
+// DialContext implements Dialer.DialContext.
+func (d *EventsSaver) DialContext(
+	ctx context.Context, network, address string) (net.Conn, error) {
+	ev := Events{Network: network, RemoteAddress: address, StartTime: time.Now()}
+	conn, err := d.Dialer.DialContext(ctx, network, address)
+	ev.EndTime = time.Now()
+	ev.Error = err
+	if conn != nil {
+		ev.LocalAddress = conn.LocalAddr().String()
+	}
+	d.mu.Lock()
+	d.events = append(d.events, ev)
+	d.mu.Unlock()
+	return conn, err
+}
+
+// Events contains events generated when dialing
+type Events struct {
+	RemoteAddress string
+	EndTime       time.Time
+	Error         error
+	LocalAddress  string
+	Network       string
+	StartTime     time.Time
+}
+
+// Service is the dialing service. This is useful to create a flexible HTTP
+// transport, for which the dialer is undertermined. Just pass an instance of
+// this struct to httptransport.NewBase(). Every dial request will block and
+// possibly timeout, until a single goroutine takes ownership of this instance
+// of Service by running service.Loop. Because service.Loop takes in input
+// a dialer, it means you can customize dial depending on circumstances.
```

Comment:
This service should help to rather easily (and explicitly) solve https://github.com/ooni/probe-engine/issues/125

Reply:
However, it seems that the solution using a bytecounter is more idiomatic Go and easy to reason about than this one, albeit I admit I quite like the idea of returning a dialer using a channel :^)